torch/distributed/fsdp/_state_dict_utils.py

```diff
@@ -696,3 +704,57 @@ def _post_load_state_dict_hook(module: nn.Module, *args: Any) -> None:
     # Dispatch into state_dict type specific implementation of post-hook for
     # loading state_dict.
     _post_load_state_dict_hook_fn[fsdp_state._state_dict_type](module, fsdp_state)
+
+
+@no_type_check
+def _register_state_dict_pre_hooks(state: _FSDPState) -> None:
+    """Registers the pre-save state dict hooks."""
+    if not _is_composable(state):
+        state.register_state_dict_pre_hook(
+            functools.partial(_pre_state_dict_hook, state)
+        )
+        return
+    for handle in state._handles:
+        handle._comm_module.register_state_dict_pre_hook(
+            functools.partial(_pre_state_dict_hook, state)
+        )
+
+
+@no_type_check
+def _register_state_dict_hooks(state: _FSDPState) -> None:
+    """Registers the post-save state dict hooks."""
+    if not _is_composable(state):
+        state._register_state_dict_hook(functools.partial(_post_state_dict_hook, state))
+        return
+    for handle in state._handles:
+        handle._comm_module._register_state_dict_hook(
+            functools.partial(_post_state_dict_hook, state)
+        )
+
+
+@no_type_check
+def _register_load_state_dict_pre_hooks(state: _FSDPState) -> None:
+    """Registers the pre-load state dict hooks."""
+    if not _is_composable(state):
+        state._register_load_state_dict_pre_hook(
+            functools.partial(_pre_load_state_dict_hook, state), with_module=True
+        )
+        return
+    for handle in state._handles:
+        handle._comm_module._register_load_state_dict_pre_hook(
+            functools.partial(_pre_load_state_dict_hook, state), with_module=True
+        )
+
+
+@no_type_check
+def _register_load_state_dict_post_hooks(state: _FSDPState) -> None:
```

Comment:
I wonder if we can generalize all this to reduce code duplication:


```
def register_state_dict_hook(register_hook_fn, hook, state, hook_fn_kwargs):
    # same code for composable vs not composable
    getattr(state, register_hook_fn)(partial(hook, state)
```

and then pass in the different things into this function.

Reply:
Done in follow-up.